### PR TITLE
TRB-43347 Disable scaling for eks spot and windows instance

### PR DESCRIPTION
**Pull request by alvinc78**
# Intent

Disable scaling for eks spot instance
Disable scaling for windows instance

# Background

We don't support these instances to be suspended or provisioned but the UI shows the actions, we want to disable that for the nodes.

# Testing

# Checklist

These are the items that must be done by the developer and by reviewers before the change is ready to merge. Please ~~strikeout~~ any items that are not applicable, but don't delete them

- [ ] Developer Checks
    - [ ] Full build with unit tests and fmt and vet checks
    - [ ] Unit tests added / updated
    - [ ] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Integration tests added / updated
    - [ ] Manual testing done (and described)
    - [ ] Product sweep run and passed
    - [ ] Developer wiki updated (and linked to this description)
- [ ] Reviewer Checks
    - [ ] Merge request description clear and understandable
    - [ ] Developer checklist items complete
    - [ ] Functional code review (how is the code written)
    - [ ] Architectural review (does the code try to do the right thing, in the right way)
    - [ ] Defensive coding (incoming data checked / sanitized, exceptions logged, clear error messages)
    - [ ] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Security review checklist complete.

# Audience

### CODE REVIEW

File: pkg/discovery/dtofactory/node_entity_dto_builder.go [hunk 1]
```diff
@@ -123,6 +123,23 @@ func (builder *nodeEntityDTOBuilder) BuildEntityDTOs(nodes []*api.Node, nodesPod
...
+			spot := strings.Contains(property.GetName(), util.EKSCapacityType) && property.GetValue() == util.EKSSpot
+			windows := strings.Contains(property.GetName(), util.NodeLabelOS) && property.GetValue() == util.WindowsOS
+			if spot || windows {
+				if spot{
+					nodeType = util.EKSSpot
+				} else if windows {
+					nodeType = util.WindowsOS
+				} else {
+					nodeType = "unknown"
+				}
```
Comment (ading1977) [inline pkg/discovery/dtofactory/node_entity_dto_builder.go:137]: I don't understand what "unknown" type means? It is possible that the node is both a spot instance and with windows OS, we just need to print the information out.

File: pkg/discovery/dtofactory/node_entity_dto_builder.go [hunk 2]
```diff
@@ -123,6 +123,23 @@ func (builder *nodeEntityDTOBuilder) BuildEntityDTOs(nodes []*api.Node, nodesPod
...
+ 				glog.Warningf("Node %s is a %s type, setting to not provisionable and not suspendable.", node.GetName(), nodeType)
```
Comment (ading1977) [inline pkg/discovery/dtofactory/node_entity_dto_builder.go:138]: This should be a v2 Info level message. Just say something like
`Suspend and provision is disabled for node that is either AWS EC2 spot instance or node with Windows OS.`

Comment (alvinc78) [reply pkg/discovery/dtofactory/node_entity_dto_builder.go]: Updated the message, thanks

File: pkg/discovery/dtofactory/node_entity_dto_builder.go [hunk 3]
```diff
@@ -123,6 +123,15 @@ func (builder *nodeEntityDTOBuilder) BuildEntityDTOs(nodes []*api.Node, nodesPod
 			glog.Errorf("Failed to get node properties: %s", err)
 			nodeActive = false
 		}
+		for _, property := range properties{
+			spot := strings.Contains(property.GetName(), util.EKSCapacityType) && property.GetValue() == util.EKSSpot
+			windows := strings.Contains(property.GetName(), util.NodeLabelOS) && property.GetValue() == util.WindowsOS
+			if spot || windows {
```
Comment (ading1977) [inline pkg/discovery/dtofactory/node_entity_dto_builder.go:129]: We can be specific on the message.

**Comment by alvinc78** [comment]
> Are you going to add a unit test here
> 
> https://github.com/turbonomic/kubeturbo/blob/92f550541093aeb3d13f2de2609055668c3efb81/pkg/discovery/dtofactory/node_entity_dto_builder_test.go#L208
> 
> seeing how complex this is, it maybe simpler to separate the extra logic into its own function and make a unit test for the small function.

I have moved this to a function itself

File: pkg/discovery/dtofactory/node_entity_dto_builder.go (hunk 3 above)
Comment (alvinc78) [reply pkg/discovery/dtofactory/node_entity_dto_builder.go]: Updated the message

**Comment by alvinc78** [comment]
> Are you going to add a unit test here
> 
> https://github.com/turbonomic/kubeturbo/blob/92f550541093aeb3d13f2de2609055668c3efb81/pkg/discovery/dtofactory/node_entity_dto_builder_test.go#L208
> 
> seeing how complex this is, it maybe simpler to separate the extra logic into its own function and make a unit test for the small function.

added unit test

File: pkg/discovery/dtofactory/node_entity_dto_builder_test.go [hunk 4]
```diff
@@ -283,3 +280,75 @@ func Test_getAffinityCommoditiesSold(t *testing.T) {
 	assert.Equal(t, proto.CommodityDTO_LABEL, *commodities[0].CommodityType)
 	assert.Equal(t, pod, *commodities[0].Key)
 }
+
+func Test_getSuspendProvisionSettingByNodeType(t *testing.T) {
+	node1 := mockNode()
+	labels1 := make(map[string]string)
+	labels1["[k8s label] beta.kubernetes.io/os"] = "windows"
```
Comment (tian-ma) [inline pkg/discovery/dtofactory/node_entity_dto_builder_test.go:287]:
```go
labels1 := map[string]string{"[k8s label] beta.kubernetes.io/os": "windows"}
```

File: pkg/discovery/dtofactory/node_entity_dto_builder_test.go [hunk 5]
```diff
@@ -283,3 +280,75 @@ func Test_getAffinityCommoditiesSold(t *testing.T) {
...
+	node1.SetLabels(labels1)
+	stitchingManager1 := stitching.NewStitchingManager(stitching.UUID)
+	stitchingManager1.StoreStitchingValue(node1)
+	nodeEntityDTOBuilder1 := NewNodeEntityDTOBuilder(metricsSink, stitchingManager1)
+	properties1, _ := nodeEntityDTOBuilder1.getNodeProperties(node1)
+
+	node2 := mockNode()
+	labels2 := make(map[string]string)
```
Comment (tian-ma) [inline pkg/discovery/dtofactory/node_entity_dto_builder_test.go:295]: no need for `make(map)`
see previous comment

**Review by tian-ma — APPROVED** [review]

File: pkg/discovery/dtofactory/node_entity_dto_builder_test.go (hunk 4 above)
Comment (alvinc78) [reply pkg/discovery/dtofactory/node_entity_dto_builder_test.go]: done

File: pkg/discovery/dtofactory/node_entity_dto_builder_test.go (hunk 5 above)
Comment (alvinc78) [reply pkg/discovery/dtofactory/node_entity_dto_builder_test.go]: done

File: pkg/discovery/dtofactory/node_entity_dto_builder.go [hunk 6]
```diff
@@ -488,3 +494,21 @@ func getNodeIPs(node *api.Node) []string {
...
+				nodeType = "AWS EC2 spot instance"
+			} else if windows {
+				nodeType = "node with Windows OS"
+			}
+			disableSuspendProvision = true
+			return
+		}
+	}
+	return
+}
```
Comment (tian-ma) [inline pkg/discovery/dtofactory/node_entity_dto_builder.go:514]: missing new line at file end is causing a build failure

File: pkg/discovery/util/node_util.go [hunk 7]
```diff
@@ -37,6 +37,10 @@ const (
 	NodePoolGKE = "cloud.google.com/gke-nodepool"
 	// EKS Node pool label
 	NodePoolEKSIdentifier = "/nodegroup" //alpha.eksctl.io/nodegroup-name or eks.amazonaws.com/nodegroup
+
+	EKSCapacityType = "eks.amazonaws.com/capacityType"
+	EKSSpot = "SPOT"
+	WindowsOS= "windows"
```
Comment (tian-ma) [inline pkg/discovery/util/node_util.go:43]: spacing is causing a build failure.
```golang
	EKSCapacityType = "eks.amazonaws.com/capacityType"
	EKSSpot         = "SPOT"
	WindowsOS       = "windows"
```

**Review by tian-ma — CHANGES_REQUESTED** [review]

File: pkg/discovery/util/node_util.go (hunk 7 above)
Comment (alvinc78) [reply pkg/discovery/util/node_util.go]: done

File: pkg/discovery/dtofactory/node_entity_dto_builder.go (hunk 6 above)
Comment (alvinc78) [reply pkg/discovery/dtofactory/node_entity_dto_builder.go]: done

**Review by tian-ma — APPROVED** [review]

**Pull request merged**